python/ray/data/_internal/execution/operators/hash_shuffle.py

```diff
@@ -1228,6 +1230,9 @@ def __init__(
                 )
             ),
             shuffle_progress_bar_name="Shuffle",
+            # NOTE: This is set to True because num_partitions (aka, # of output blocks)
+            # must be preserved.
+            disallow_block_splitting=True,
```

Comment:
```suggestion
            # NOTE: In cases like ``groupby`` blocks can't be split as this might violate an invariant that all rows
            #             with the same key are in the same group (block)
            disallow_block_splitting=True,
```